### ISSUE TITLE
pulumi -r causes recreation of ElastiCache resources every single time

### ISSUE THREAD

**Issue by armenr** [pulumi/pulumi-aws#684]
Every time I run `pulumi up -r `, the following resources are all recreated/replaced.

Can someone tell me what I might be doing wrong to be causing this?

The VPC I'm working in has 1 isolated subnet, and spans 2 Availability Zones.

I don't get why it wants to recreate on every single refresh. 

### Code below:

```
// Assign "isolated" redis Subnet to redisCluster
const redisClusterSubnets = new aws.elasticache.SubnetGroup(nameFactory('sessionSubnets'), {
  subnetIds: projectVPC.isolatedSubnetIds,
});

const redisClusterAccess = new awsx.ec2.SecurityGroup(nameFactory('redis-webhosts-access'), {
  vpc: projectVPC,
  ingress: [redisAllowEC2Rule],
  egress: [egressToAllRule],
  description: 'Redis Security Group',
});

// Create redis Cluster (via Replication Group)
const redisCluster = new aws.elasticache.ReplicationGroup(nameFactory('sessions'), {
  // https://pulumi.io/reference/pkg/nodejs/pulumi/aws/elasticache/#ReplicationGroup
  // replicationGroupId: nameFactory('sessions'),
  availabilityZones: projectVPC.isolatedSubnets.map((item) => item.subnet.availabilityZone),
  applyImmediately: true,
  atRestEncryptionEnabled: true,
  automaticFailoverEnabled: true,
  autoMinorVersionUpgrade: true,
  clusterMode: {
    numNodeGroups: 1,
    replicasPerNodeGroup: 1,
  },
  engineVersion: '5.0.4',
  maintenanceWindow: 'wed:12:00-wed:13:00',
  nodeType: 'cache.t2.small',
  parameterGroupName: 'default.redis5.0.cluster.on',
  port: 6379,
  replicationGroupDescription: 'session storage Redis cluster',
  securityGroupIds: [redisClusterAccess.id],
  subnetGroupName: redisClusterSubnets.name,
});
```

**Comment by ellismg** [comment]
Could you let us know what version of the `pulumi` CLI you are using as well as what version of the `@pulumi/aws` package you have?

If you run `pulumi refresh` without any changes, I would expect that pulumi will say it wants to make changes to some resources, could you share with us the diffs that it detects?

**Comment by armenr** [comment]
Pulumi version:
```
╰─ pulumi version
v0.17.27
```

```
╰─ cat package-lock.json| grep pulumi
        "@pulumi/aws": {
            "resolved": "https://registry.npmjs.org/@pulumi/aws/-/aws-0.18.16.tgz",
                "@pulumi/pulumi": "^0.17.20",
        "@pulumi/awsx": {
            "resolved": "https://registry.npmjs.org/@pulumi/awsx/-/awsx-0.18.6.tgz",
                "@pulumi/aws": "^0.18.9",
                "@pulumi/docker": "^0.17.0",
                "@pulumi/pulumi": "^0.17.17",
        "@pulumi/docker": {
            "resolved": "https://registry.npmjs.org/@pulumi/docker/-/docker-0.17.0.tgz",
                "@pulumi/pulumi": "^v0.17.0",
        "@pulumi/pulumi": {
            "resolved": "https://registry.npmjs.org/@pulumi/pulumi/-/pulumi-0.17.21.tgz",
                "@pulumi/query": "^0.3.0",
        "@pulumi/query": {
            "resolved": "https://registry.npmjs.org/@pulumi/query/-/query-0.3.0.tgz",
```

**Note:** There's also a weird issue with security groups I'm running into, using awsx + ALB + attachment groups & specifying a default security group - so you'll notice it's trying to also replace/change that. 

I *was* going include that in a different issue :)

### Diffs:
```
  pulumi:pulumi:Stack: (same)
    [urn=urn:pulumi:dev::gruv-tessitura::pulumi:pulumi:Stack::gruv-tessitura-dev]
        > aws:ec2/securityGroup:SecurityGroup: (read)
            [id=sg-007c7bfce8936bcb2]
            [urn=urn:pulumi:dev::gruv-tessitura::awsx:x:elasticloadbalancingv2:ApplicationLoadBalancer$aws:ec2/securityGroup:SecurityGroup::dev-loadbalancer-0]
          - description        : "Default EC2 security group"
          - egress             : [
          -     [0]: {
                  - cidrBlocks    : [
                  -     [0]: "0.0.0.0/0"
                    ]
                  - description   : "Externally available at port 443"
                  - fromPort      : 443
                  - protocol      : "tcp"
                  - self          : false
                  - toPort        : 443
                }
          -     [1]: {
                  - cidrBlocks    : [
                  -     [0]: "0.0.0.0/0"
                    ]
                  - description   : "Externally available at port 80"
                  - fromPort      : 80
                  - protocol      : "tcp"
                  - self          : false
                  - toPort        : 80
                }
          -     [2]: {
                  - cidrBlocks    : [
                  -     [0]: "0.0.0.0/0"
                    ]
                  - fromPort      : 0
                  - protocol      : "-1"
                  - self          : false
                  - toPort        : 0
                }
            ]
          - ingress            : [
          -     [0]: {
                  - cidrBlocks    : [
                  -     [0]: "0.0.0.0/0"
                    ]
                  - description   : "Externally available at port 443"
                  - fromPort      : 443
                  - protocol      : "tcp"
                  - self          : false
                  - toPort        : 443
                }
          -     [1]: {
                  - cidrBlocks    : [
                  -     [0]: "0.0.0.0/0"
                    ]
                  - description   : "Externally available at port 80"
                  - fromPort      : 80
                  - protocol      : "tcp"
                  - self          : false
                  - toPort        : 80
                }
            ]
          - name               : "dev-sg-ALBPublicDefault-32ae17a"
          - revokeRulesOnDelete: false
          - tags               : {}
          - vpcId              : "vpc-03e60ecc4e8aaee96"
    ++aws:elasticache/subnetGroup:SubnetGroup: (create-replacement)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]
      ~ name       : "dev-sessionsubnets-a18a4c7" => "dev-sessionSubnets-be3c3c7"
      ~ subnetIds  : [
          ~ [0]: "subnet-0907008e0b4c6ec65" => "subnet-0ee64d676c4c93d9e"
          ~ [1]: "subnet-0ee64d676c4c93d9e" => "subnet-0907008e0b4c6ec65"
        ]
    +-aws:elasticache/subnetGroup:SubnetGroup: (replace)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]
      ~ name       : "dev-sessionsubnets-a18a4c7" => "dev-sessionSubnets-be3c3c7"
      ~ subnetIds  : [
          ~ [0]: "subnet-0907008e0b4c6ec65" => "subnet-0ee64d676c4c93d9e"
          ~ [1]: "subnet-0ee64d676c4c93d9e" => "subnet-0907008e0b4c6ec65"
        ]
    ++aws:elasticache/replicationGroup:ReplicationGroup: (create-replacement)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
      ~ availabilityZones          : [
          ~ [0]: "us-west-2b" => "us-west-2a"
          ~ [1]: "us-west-2a" => "us-west-2b"
        ]
      - numberCacheClusters        : 2
      ~ replicationGroupId         : "dev-sessions-13017c2" => "dev-sessions-e50a97d"
      - securityGroupNames         : []
      - snapshotWindow             : "07:30-08:30"
      ~ subnetGroupName            : "dev-sessionsubnets-a18a4c7" => output<string>
    +-aws:elasticache/replicationGroup:ReplicationGroup: (replace)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
      ~ availabilityZones          : [
          ~ [0]: "us-west-2b" => "us-west-2a"
          ~ [1]: "us-west-2a" => "us-west-2b"
        ]
      - numberCacheClusters        : 2
      ~ replicationGroupId         : "dev-sessions-13017c2" => "dev-sessions-e50a97d"
      - securityGroupNames         : []
      - snapshotWindow             : "07:30-08:30"
      ~ subnetGroupName            : "dev-sessionsubnets-a18a4c7" => output<string>
    ~ aws:route53/record:Record: (update)
        [id=Z194PL86ODOIXO_dev-redis_CNAME]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:route53/record:Record::dev-redis]
      ~ records: [
          ~ [0]: "dev-sessions-13017c2.buh943.clustercfg.usw2.cache.amazonaws.com" => output<string>
        ]
            --outputs:--
            arn                : "arn:aws:ec2:us-west-2:491419166139:security-group/sg-007c7bfce8936bcb2"
            description        : "Default EC2 security group"
            egress             : [
                [0]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 443"
                    fromPort      : 443
                    protocol      : "tcp"
                    self          : false
                    toPort        : 443
                }
                [1]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 80"
                    fromPort      : 80
                    protocol      : "tcp"
                    self          : false
                    toPort        : 80
                }
                [2]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    fromPort      : 0
                    protocol      : "-1"
                    self          : false
                    toPort        : 0
                }
            ]
            id                 : "sg-007c7bfce8936bcb2"
            ingress            : [
                [0]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 443"
                    fromPort      : 443
                    protocol      : "tcp"
                    self          : false
                    toPort        : 443
                }
                [1]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 80"
                    fromPort      : 80
                    protocol      : "tcp"
                    self          : false
                    toPort        : 80
                }
            ]
            name               : "dev-sg-ALBPublicDefault-32ae17a"
            ownerId            : "491419166139"
          - revokeRulesOnDelete: false
            vpcId              : "vpc-03e60ecc4e8aaee96"
    --aws:elasticache/replicationGroup:ReplicationGroup: (delete-replaced)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
    --aws:elasticache/subnetGroup:SubnetGroup: (delete-replaced)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::gruv-tessitura::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]


```

**Comment by armenr** [comment]
Pulumi Version: 
```
╰─ pulumi version
v0.17.27
```

Packages:
```
        "@pulumi/aws": {
            "resolved": "https://registry.npmjs.org/@pulumi/aws/-/aws-0.18.16.tgz",
                "@pulumi/pulumi": "^0.17.20",
        "@pulumi/awsx": {
            "resolved": "https://registry.npmjs.org/@pulumi/awsx/-/awsx-0.18.6.tgz",
                "@pulumi/aws": "^0.18.9",
                "@pulumi/docker": "^0.17.0",
                "@pulumi/pulumi": "^0.17.17",
        "@pulumi/docker": {
            "resolved": "https://registry.npmjs.org/@pulumi/docker/-/docker-0.17.0.tgz",
                "@pulumi/pulumi": "^v0.17.0",
        "@pulumi/pulumi": {
            "resolved": "https://registry.npmjs.org/@pulumi/pulumi/-/pulumi-0.17.21.tgz",
                "@pulumi/query": "^0.3.0",
        "@pulumi/query": {
            "resolved": "https://registry.npmjs.org/@pulumi/query/-/query-0.3.0.tgz",
```

## Diff:

```
    ++aws:elasticache/subnetGroup:SubnetGroup: (create-replacement)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::project-sample::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]
      ~ name       : "dev-sessionsubnets-a18a4c7" => "dev-sessionSubnets-82e8045"
      ~ subnetIds  : [
          ~ [0]: "subnet-0907008e0b4c6ec65" => "subnet-0ee64d676c4c93d9e"
          ~ [1]: "subnet-0ee64d676c4c93d9e" => "subnet-0907008e0b4c6ec65"
        ]
    +-aws:elasticache/subnetGroup:SubnetGroup: (replace)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::project-sample::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]
      ~ name       : "dev-sessionsubnets-a18a4c7" => "dev-sessionSubnets-82e8045"
      ~ subnetIds  : [
          ~ [0]: "subnet-0907008e0b4c6ec65" => "subnet-0ee64d676c4c93d9e"
          ~ [1]: "subnet-0ee64d676c4c93d9e" => "subnet-0907008e0b4c6ec65"
        ]
    ++aws:elasticache/replicationGroup:ReplicationGroup: (create-replacement)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::project-::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
      ~ availabilityZones          : [
          ~ [0]: "us-west-2b" => "us-west-2a"
          ~ [1]: "us-west-2a" => "us-west-2b"
        ]
      - numberCacheClusters        : 2
      ~ replicationGroupId         : "dev-sessions-13017c2" => "dev-sessions-cbd87e3"
      - securityGroupNames         : []
      - snapshotWindow             : "07:30-08:30"
      ~ subnetGroupName            : "dev-sessionsubnets-a18a4c7" => output<string>
    +-aws:elasticache/replicationGroup:ReplicationGroup: (replace)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::project-sample::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
      ~ availabilityZones          : [
          ~ [0]: "us-west-2b" => "us-west-2a"
          ~ [1]: "us-west-2a" => "us-west-2b"
        ]
      - numberCacheClusters        : 2
      ~ replicationGroupId         : "dev-sessions-13017c2" => "dev-sessions-cbd87e3"
      - securityGroupNames         : []
      - snapshotWindow             : "07:30-08:30"
      ~ subnetGroupName            : "dev-sessionsubnets-a18a4c7" => output<string>
    ~ aws:route53/record:Record: (update)
        [id=Z194PL86ODOIXO_dev-redis_CNAME]
        [urn=urn:pulumi:dev::project-sample::aws:route53/record:Record::dev-redis]
      ~ records: [
          ~ [0]: "dev-sessions-13017c2.buh943.clustercfg.usw2.cache.amazonaws.com" => output<string>
        ]
            --outputs:--
            arn                : "arn:aws:ec2:us-west-2:491419166139:security-group/sg-007c7bfce8936bcb2"
            description        : "Default EC2 security group"
            egress             : [
                [0]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 443"
                    fromPort      : 443
                    protocol      : "tcp"
                    self          : false
                    toPort        : 443
                }
                [1]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 80"
                    fromPort      : 80
                    protocol      : "tcp"
                    self          : false
                    toPort        : 80
                }
                [2]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    fromPort      : 0
                    protocol      : "-1"
                    self          : false
                    toPort        : 0
                }
            ]
            id                 : "sg-007c7bfce8936bcb2"
            ingress            : [
                [0]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 443"
                    fromPort      : 443
                    protocol      : "tcp"
                    self          : false
                    toPort        : 443
                }
                [1]: {
                    cidrBlocks    : [
                        [0]: "0.0.0.0/0"
                    ]
                    description   : "Externally available at port 80"
                    fromPort      : 80
                    protocol      : "tcp"
                    self          : false
                    toPort        : 80
                }
            ]
            name               : "dev-sg-ALBPublicDefault-32ae17a"
            ownerId            : "REDACTED"
          - revokeRulesOnDelete: false
            vpcId              : "vpc-03e60ecc4e8aaee96"
    --aws:elasticache/replicationGroup:ReplicationGroup: (delete-replaced)
        [id=dev-sessions-13017c2]
        [urn=urn:pulumi:dev::project-sample::aws:elasticache/replicationGroup:ReplicationGroup::dev-sessions]
    --aws:elasticache/subnetGroup:SubnetGroup: (delete-replaced)
        [id=dev-sessionsubnets-a18a4c7]
        [urn=urn:pulumi:dev::project-sample::aws:elasticache/subnetGroup:SubnetGroup::dev-sessionSubnets]

```

**Comment by armenr** [comment]
I updated all my packages. Still getting the same thing. By the looks of the referenced commits, I'm guessing there was an issue?

**Comment by armenr** [comment]
Hey there @pgavlin - Thanks for addressing the issue and getting the bug patched...but is there any way I can test this out? I don't see a new tagged release, so just wondering when we can get the benefit of this fix. 

Thanks again!

**Comment by pgavlin** [comment]
Hi @armenr! We'll be working on getting the fix rolled out over the next few days.

**Comment by pgavlin** [comment]
We are still working on releasing the next version of `@pulumi/aws`, which will contain this fix. Thanks for your patience.

**Comment by armenr** [comment]
Hey there @pgavlin - Just checking in to see how this is going? I'm still having the same issue - I've updated to the latest releases of pulumi (command-line) + the latest tagged releases of the corresponding nodeJS packages. 

```

**Comment by pgavlin** [comment]
Okay, so I've verified that we've rolled out the fix with v0.18.27 of the Node package, but I've also been able to reproduce the behavior you're seeing. The problem here is that we perform the `refresh` using the _old_ version of the AWS provider, which still contains the original bug. You can migrate your stack to the new version of the AWS provider by running a `pulumi up` without the `-r` flag. Once you've done this, your resources should all reference the newer version of the AWS provider and you should no longer see the spurious replacement.
